plugin/src/main/java/at/helpch/chatchat/placeholder/MiniPlaceholderImpl.java

```diff
@@ -42,34 +44,40 @@ public MiniPlaceholderImpl(
         this.message = message;
     }
 
-    public @NotNull TagResolver toTagResolver(
-        final boolean inMessage,
-        @NotNull final ChatUser sender,
-        @NotNull User recipient
-    ) {
-        if (inMessage && !sender.player().hasPermission(MINI_PLACEHOLDER_PERMISSION + name)) {
+    public @NotNull TagResolver toTagResolver(final @NotNull Context context) {
+        if (context.inMessage() && context.sender().isPresent() && !context.sender().get().player().hasPermission(MINI_PLACEHOLDER_PERMISSION + name)) {
             return TagResolver.empty();
         }
 
         if (!parseMini) {
             return Placeholder.unparsed(name, message);
         }
 
-        final TagResolver papiTag;
-        if (parsePapi) {
-            if (requiresRecipient && recipient instanceof ChatUser) {
-                papiTag = TagResolver.resolver(
-                    PapiTagUtils.createPlaceholderAPITag(sender.player()),
-                    PapiTagUtils.createRelPlaceholderAPITag(sender.player(), ((ChatUser) recipient).player()),
-                    PapiTagUtils.createRecipientTag(((ChatUser) recipient).player())
-                );
-            } else {
-                papiTag = PapiTagUtils.createPlaceholderAPITag(sender.player());
-            }
-        } else {
-            papiTag = TagResolver.empty();
+        final Optional<ChatUser> sender = context.sender();
+        final Optional<User> recipient = context.recipient();
+
+        if (sender.isEmpty()) {
+            return TagResolver.empty();
+        }
+
+        if (parsePapi && requiresRecipient && recipient.isEmpty()) {
+            return TagResolver.empty();
         }
 
+        if (!parsePapi) {
+            return closing
+                ? Placeholder.component(name, MessageUtils.parseToMiniMessage(message))
+                : TagResolver.resolver(name, Tag.inserting(MessageUtils.parseToMiniMessage(message)));
+        }
+
+        final TagResolver papiTag = !requiresRecipient || !(recipient.get() instanceof ChatUser)
+            ? PapiTagUtils.createPlaceholderAPITag(sender.get().player())
+            : TagResolver.resolver(
+                PapiTagUtils.createPlaceholderAPITag(sender.get().player()),
+                PapiTagUtils.createRelPlaceholderAPITag(sender.get().player(), ((ChatUser) recipient.get()).player()),
+                PapiTagUtils.createRecipientTag(((ChatUser) recipient.get()).player())
+            );
```

Comment:
Is possible to improve this? Is kinda hard to read

Reply:
I've improved the variable naming and also the logic order to hopefully make it a lot easier to read.